File: packages/core/src/dependencies/DependencyGraph.ts
import type { PathSegment } from '../expressions/ExpressionNode.js';
import type { DependencyError } from './DependencyError.js';
import type { DependencyNode } from './DependencyNode.js';
import { pathMatchesPattern } from './DependencyPattern.js';

export interface DependencyPlan {
  /** Nodes to recompute, dependencies before dependents. */
  readonly order: readonly string[];
  readonly errors: readonly DependencyError[];
}

/**
 * Which nodes must re-evaluate when a value changes, and in what order.
 *
 * Push-based and synchronous by design (ADR-0004): no scheduler, no microtask
 * batching. An adapter that wants frame-level batching does that at the adapter, so
 * core stays deterministic and its tests stay order-independent.
 */
export class DependencyGraph {
  readonly #nodes: Map<string, DependencyNode> = new Map();

  addNode(node: DependencyNode): void {
    if (this.#nodes.has(node.key)) {
      throw new Error(`Dependency node "${node.key}" is already registered.`);
    }
    this.#nodes.set(node.key, node);
  }

  /** Replaces a registration, or adds one. Re-registering an edited expression. */
  setNode(node: DependencyNode): void {
    this.#nodes.set(node.key, node);
  }

  removeNode(key: string): void {
    this.#nodes.delete(key);
  }

  hasNode(key: string): boolean {
    return this.#nodes.has(key);
  }

  /** Sorted, so every derived ordering is stable across runs. */
  getNodeKeys(): readonly string[] {
    return Array.from(this.#nodes.keys()).toSorted();
  }

  /** Nodes that read something a change at `path` touches. */
  dependentsOf(path: readonly PathSegment[]): readonly string[] {
    return this.getNodeKeys().filter((key) => {
      const node = this.#nodes.get(key);
      return node !== undefined && node.reads.some((pattern) => pathMatchesPattern(path, pattern));
    });
  }

  /**
   * Everything a set of changes affects, in dependency order.
   *
   * Expansion follows declared writes, so a calculated value feeding another is picked
   * up here rather than by re-running until the values settle.
   */
  plan(changedPaths: readonly (readonly PathSegment[])[]): DependencyPlan {
    const affected = this.#collectAffected(changedPaths);
    return this.#order(affected);
  }

  /**
   * Every node, in dependency order.
   *
   * Used for the initial evaluation, where there is no change to propagate from.
   * Ordering still matters there: a rule that writes a value another rule reads has to
   * run first, or the second sees a stale answer on the very first pass. It also
   * surfaces cycles at load time rather than on the first keystroke.
   */
  planAll(): DependencyPlan {
    return this.#order(new Set(this.getNodeKeys()));
  }

  #collectAffected(changedPaths: readonly (readonly PathSegment[])[]): Set<string> {
    const affected = new Set<string>();
    const queue: string[] = [];

    for (const path of changedPaths) {
      for (const key of this.dependentsOf(path)) {
        if (!affected.has(key)) {
          affected.add(key);
          queue.push(key);
        }
      }
    }

    while (queue.length > 0) {
      const key = queue.shift();
      const written = key === undefined ? undefined : this.#nodes.get(key)?.writes;
      if (written === undefined) {
        continue;
      }
      for (const dependent of this.dependentsOf(written)) {
        if (!affected.has(dependent)) {
          affected.add(dependent);
          queue.push(dependent);
        }
      }
    }

    return affected;
  }

  /**
   * Nodes inside `affected` whose writes feed this node's reads.
   *
   * A node is deliberately allowed to be its own predecessor. A calculated value that
   * reads what it writes — `total = total + 1` — is a cycle of one, and excluding self
   * here would let the single most obvious authoring mistake through undetected.
   */
  #predecessorsOf(key: string, affected: ReadonlySet<string>): readonly string[] {
    const node = this.#nodes.get(key);
    if (node === undefined) {
      return [];
    }
    return this.getNodeKeys().filter((candidate) => {
      if (!affected.has(candidate)) {
        return false;
      }
      const written = this.#nodes.get(candidate)?.writes;
      return (
        written !== undefined && node.reads.some((pattern) => pathMatchesPattern(written, pattern))
      );
    });
  }

  /**
   * Depth-first topological sort.
   *
   * DFS rather than Kahn's algorithm specifically so a cycle can be reported as the
   * path that closes it: the visiting stack *is* the cycle when a back edge is found.
   */
  #order(affected: ReadonlySet<string>): DependencyPlan {
    const order: string[] = [];
    const errors: DependencyError[] = [];
    const state = new Map<string, 'visiting' | 'done'>();
    const stack: string[] = [];

    const visit = (key: string): void => {
      const current = state.get(key);
      if (current === 'done') {
        return;
      }
      if (current === 'visiting') {
        const start = stack.indexOf(key);
        const cycle = [...stack.slice(start === -1 ? 0 : start), key];
        errors.push({
          code: 'cycle',
          message: `Dependency cycle: ${cycle.join(' -> ')}.`,
          nodes: cycle,
        });
        return;
      }

      state.set(key, 'visiting');
      stack.push(key);
      for (const predecessor of this.#predecessorsOf(key, affected)) {
        visit(predecessor);
      }
      stack.pop();
      state.set(key, 'done');
      order.push(key);
    };

    for (const key of this.getNodeKeys()) {
      if (affected.has(key)) {
        visit(key);
      }
    }

    return { order, errors };
  }
}

File: dotnet/src/Kajay.Core/DependencyGraph.cs
namespace Kajay;

internal sealed class DependencyGraph
{
    private readonly Dictionary<string, DependencyNode> _nodes = new(StringComparer.Ordinal);
    private readonly Dictionary<string, string[]> _outgoing = new(StringComparer.Ordinal);
    private readonly Dictionary<string, string[]> _predecessors = new(StringComparer.Ordinal);
    private bool _edgesAreCurrent;

    internal void AddNode(DependencyNode node)
    {
        ArgumentNullException.ThrowIfNull(node);
        if (!_nodes.TryAdd(node.Key, node))
        {
            throw new ArgumentException(
                $"Dependency node '{node.Key}' is already registered.",
                nameof(node));
        }

        _edgesAreCurrent = false;
    }

    internal void SetNode(DependencyNode node)
    {
        ArgumentNullException.ThrowIfNull(node);
        _nodes[node.Key] = node;
        _edgesAreCurrent = false;
    }

    internal bool RemoveNode(string key)
    {
        ArgumentNullException.ThrowIfNull(key);
        bool removed = _nodes.Remove(key);
        _edgesAreCurrent &= !removed;
        return removed;
    }

    internal DependencyPlan Plan(IReadOnlyList<ExpressionPath> changedPaths)
    {
        ArgumentNullException.ThrowIfNull(changedPaths);
        return Order(CollectAffected(changedPaths));
    }

    internal DependencyPlan PlanAll()
    {
        return Order(new HashSet<string>(_nodes.Keys, StringComparer.Ordinal));
    }

    private HashSet<string> CollectAffected(IReadOnlyList<ExpressionPath> changedPaths)
    {
        EnsureEdges();
        var affected = new HashSet<string>(StringComparer.Ordinal);
        var queue = new Queue<string>();
        foreach (ExpressionPath path in changedPaths)
        {
            AddDependents(path, affected, queue);
        }

        while (queue.TryDequeue(out string? key))
        {
            foreach (string dependent in _outgoing[key])
            {
                if (affected.Add(dependent))
                {
                    queue.Enqueue(dependent);
                }
            }
        }

        return affected;
    }

    private void AddDependents(
        ExpressionPath path,
        HashSet<string> affected,
        Queue<string> queue)
    {
        foreach (string key in SortedKeys())
        {
            DependencyNode node = _nodes[key];
            if (node.Reads.Any(pattern => pattern.Overlaps(path)) && affected.Add(key))
            {
                queue.Enqueue(key);
            }
        }
    }

    private DependencyPlan Order(HashSet<string> affected)
    {
        EnsureEdges();
        List<string> order = [];
        List<DependencyError> errors = [];
        var states = new Dictionary<string, VisitState>(StringComparer.Ordinal);
        List<string> stack = [];

        void Visit(string key)
        {
            if (states.GetValueOrDefault(key) is VisitState.Done)
            {
                return;
            }

            if (states.GetValueOrDefault(key) is VisitState.Visiting)
            {
                int start = stack.IndexOf(key);
                string[] cycle = [.. stack.Skip(start < 0 ? 0 : start), key];
                errors.Add(new DependencyError("cycle", cycle));
                return;
            }

            states[key] = VisitState.Visiting;
            stack.Add(key);
            foreach (string predecessor in PredecessorsOf(key, affected))
            {
                Visit(predecessor);
            }

            stack.RemoveAt(stack.Count - 1);
            states[key] = VisitState.Done;
            order.Add(key);
        }

        foreach (string key in SortedKeys())
        {
            if (affected.Contains(key))
            {
                Visit(key);
            }
        }

        return new DependencyPlan(order.ToArray(), errors.ToArray());
    }

    private string[] PredecessorsOf(
        string key,
        HashSet<string> affected)
    {
        return _predecessors[key]
            .Where(affected.Contains)
            .ToArray();
    }

    private void EnsureEdges()
    {
        if (_edgesAreCurrent)
        {
            return;
        }

        _outgoing.Clear();
        _predecessors.Clear();
        string[] keys = SortedKeys();
        foreach (string key in keys)
        {
            _outgoing[key] = [];
            _predecessors[key] = [];
        }

        foreach (string writerKey in keys)
        {
            if (_nodes[writerKey].Writes is not ExpressionPath written)
            {
                continue;
            }

            string[] dependents = keys
                .Where(dependent => _nodes[dependent].Reads.Any(
                    pattern => pattern.Overlaps(written)))
                .ToArray();
            _outgoing[writerKey] = dependents;
            foreach (string dependent in dependents)
            {
                _predecessors[dependent] = [.. _predecessors[dependent], writerKey];
            }
        }

        _edgesAreCurrent = true;
    }

    private string[] SortedKeys()
    {
        return _nodes.Keys.Order(StringComparer.Ordinal).ToArray();
    }

    private enum VisitState
    {
        Unvisited,
        Visiting,
        Done,
    }
}
